### PR TITLE
Create a generic archive download manager

### PR DESCRIPTION
**Pull request by naoey**
Moves the beatmap download logic out of `BeatmapManager` into a base class that can be shared by other model managers that need to implement download functionality.

As per the discussion in #4985 this is the first of multiple PRs to centralise model downloading logic.

### CODE REVIEW

File: osu.Game/Database/ArchiveModelManager.cs
Comment (peppy) [inline osu.Game/Database/ArchiveModelManager.cs]: intentional change? if so, `ItemAddedDelegate` should be removed.

File: osu.Game/Database/IModelManager.cs
Comment (peppy) [inline osu.Game/Database/IModelManager.cs]: needs xmldoc. also this doesn't contain the correct constraints on `TModel`.

also this isn't really used anywhere. not sure we should be defining such interfaces too early.

File: osu.Game/Database/IModelDownloader.cs
Comment (peppy) [inline osu.Game/Database/IModelDownloader.cs]: (class) xmldoc should be placed on interface instead of abstract class, if we decide to go in this direction.

**Review by peppy — CHANGES_REQUESTED** [review]

File: osu.Game/Database/IModelManager.cs
Comment (naoey) [reply osu.Game/Database/IModelManager.cs]: agree that it was early to include in this PR. it is used in the `DownloadTrackingComposite` update PR (#5001) to consume the items added/removed events.

also was (am still) unsure what the right constraints to have here at the interface level would be.

Comment (peppy) [reply osu.Game/Database/IModelManager.cs]: Ah, I didn't realise that uses the interface. Is that for DI purposes?

Comment (naoey) [reply osu.Game/Database/IModelManager.cs]: well its not used directly but by virtue of `IModelDownloader` inheriting it (which is used for the DI). the `ItemAdded`/`ItemRemoved` events are needed on the generic type.

Comment (peppy) [reply osu.Game/Database/IModelManager.cs]: Right. Understood.

File: osu.Game/Online/API/ArchiveDownloadRequest.cs
Comment (peppy) [inline osu.Game/Online/API/ArchiveDownloadRequest.cs]: Rename this to Model

**Comment by peppy** [comment]
I made some changes to simplify the download logic (since we don't really use non-typed objects anywhere in osu!). Please check you agree with this and also that it works OK for your future PRs.

**Comment by naoey** [comment]
I had initially implemented something like this, but I considered that a binary value option may be too limiting for some downloads in the future. Of course this depends on systems that don’t exist right now, so I’ll leave it to you to decide which is better based on the direction you’d want it to take. Or whether to revisit this at that point. 

Eg. I was thinking that with something like a skin download, we might want to use multiple parameters like which game mode(s), whether to include extras like cursor and hitsound packs etc.

Other than that this will work as is with my next PRs 👍

**Comment by peppy** [comment]
If we need more than just a `bool`, the correct direction would be to make a generic DownloadRequest object which can be derived:

```
interface DownloadRequest<T>
{
    T Model { get; set; }
}

BeatmapDownloadRequest : DownloadRequest<BeatmapSetInfo>
{
    BeatmapSetInfo Model { get; set; }

    public bool NoVideo { get; set; }
}
```

**Comment by naoey** [comment]
But `CreateDownloadRequest` would need the parameters to initialise the DownloadRequest object correctly?

**Comment by peppy** [comment]
it would become:

`CreateDownloadRequest(DownloadRequest<TModel> request)`

**Comment by naoey** [comment]
That seems a bit backwards. `CreateDownloadRequest` should take a `TModel` argument and return a `DownloadRequest<TModel>`, but here it takes the `DownloadRequest<TModel>` as the argument?

In that case can't we just expose the `performDownloadWithRequest` private method to actually do the download, and externalise the `DownloadRequest` creation completely?

**Comment by peppy** [comment]
Oh sorry, I mistyped.

```csharp
public Download(DownloadRequest<TModel> request);

protected override ArchiveDownloadRequest<BeatmapSetInfo> CreateDownloadRequest(DownloadRequest<TModel> request);
```

The new proposed `DownloadRequest` would be the input to the download system (from the `DownloadButton` etc.). It could be made a secondary overload to allow for still making simple downlaods via `Download(TModel model)` still.

**Comment by naoey** [comment]
Ah I see now. Yes that would be better.

**Comment by peppy** [comment]
Best not over-complicate until necessary, so (optional) bool is fine for the time being.

File: osu.Game/Online/API/Requests/DownloadBeatmapSetRequest.cs
Comment (naoey) [inline osu.Game/Online/API/Requests/DownloadBeatmapSetRequest.cs]: just noticed this unnecessary local variable. `Model` is already on parent class.

Comment (peppy) [reply osu.Game/Online/API/Requests/DownloadBeatmapSetRequest.cs]: feel free to fix if you are able to. i'll be doing another pass on this later on failing that.

File: osu.Game/Database/DownloadableArchiveModelManager.cs
Comment (peppy) [inline osu.Game/Database/DownloadableArchiveModelManager.cs]: should be double-checked as to why before this is merged as a valid solution. (note to self if no one beats me).

**Comment by peppy** [comment]
@naoey you should be able to remove the try-catch logic after ppy/osu-framework#2572 is merged.

File: osu.Game/Database/ArchiveModelManager.cs
Comment (peppy) [reply osu.Game/Database/ArchiveModelManager.cs]: still not sure about this one. the reason it had a delegate is that the bool here is ambiguous.

Comment (naoey) [reply osu.Game/Database/ArchiveModelManager.cs]: there wasn't really any particular reason for changing this other than avoiding declaring a delegate in `IModelManager`. should it be added back on the interface or should we just add it to the documentation?

**Comment by peppy** [comment]
I ended up removing the `bool` as it was never used correctly and does not seem necessary.

**Review by peppy — APPROVED** [review]

**Pull request merged**
